tf_trees/tel.py

```diff
@@ -0,0 +1,190 @@
+        leading to a 2D tensor of shape=[batch_size, output_logits_dim].
+        Otherwise, the tree outputs are not added and the layer output is
+        a 2D tensor of shape=[batch_size, trees_num * output_logits_dim].
+      parallelize_over_samples: Boolean, If true, parallelizes the updates over
+        the samples in each tree. Might lead to speedups when the number of
+        trees is small (at the cost of extra memory consumption).
+      split_initializer: A Keras initializer for the internal (splitting) nodes.
+      leaf_initializer: A Keras initializer for the leaves.
+      split_regularizer: A Keras regularizer for the internal (splitting) nodes.
+      leaf_regularizer: A Keras regularizer for the leaves.
+    Input shape: A tensor of shape=[batch_size, input_dim].
+    Output shape: A tensor of shape=[batch_size, output_logits_dim] if
+      sum_outputs=True. Otherwise, a tensor of shape=[batch_size, trees_num *
+      output_logits_dim].
+    """
+
+    def __init__(self,
+                 output_logits_dim,
+                 trees_num=1,
+                 depth=3,
+                 smooth_step_param=1.0,
+                 sum_outputs=True,
+                 parallelize_over_samples=False,
+                 split_initializer=RandomUniform(-0.01, 0.01),
+                 leaf_initializer=RandomUniform(-0.01, 0.01),
+                 split_regularizer=None,
+                 leaf_regularizer=None,
+                 **kwargs):
+      """Initializes neural trees layer."""
+      self.trees_num = trees_num
+      self.depth = depth
+      self.output_logits_dim = output_logits_dim
+      self.smooth_step_param = smooth_step_param
+      self.parallelize_over_samples = parallelize_over_samples
+      self.sum_outputs = sum_outputs
+      self.split_initializer = keras.initializers.get(split_initializer)
+      self.leaf_initializer = keras.initializers.get(leaf_initializer)
+      self.split_regularizer = keras.regularizers.get(split_regularizer)
+      self.leaf_regularizer = keras.regularizers.get(leaf_regularizer)
+      super(TEL, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+      """Creates a keras layer."""
+      num_leaves = 2**self.depth
+      num_internal_nodes = num_leaves - 1
+      input_dim = input_shape[1]
+      # A list of node weights. Each element corresponds to one tree.
+      self.node_weights = []
+      # A list of leaf weights. Each element corresponds to one tree.
+      self.leaf_weights = []
+      # Create node/leaf weights for every tree.
+      for _ in range(self.trees_num):
+        self.node_weights.append(
+            self.add_weight(
+                name='node_weights',
+                shape=[input_dim, num_internal_nodes],
+                initializer=self.split_initializer,
+                trainable=True,
+                regularizer=self.split_regularizer))
+
+        self.leaf_weights.append(
+            self.add_weight(
```

Comment:
name the private methods also with _:
_add_weight

Reply:
add_weight is a method in the base class no?

Reply:
Oh, ignore me